### ISSUE TITLE
Is RAG (document grounding) feature integrated in Spring AI?

### ISSUE THREAD

**Issue by zxcvboy** [SAP/ai-sdk-java#241]
### Ask the Question

Dear colleague. 

    Is RAG feature supported by ai-sdk-java? If not, i there any roadmap? 

    Thanks a lot.

**Comment by jjtang1985** [comment]
Hi @zxcvboy ,

Could you please share more details about your use cases?
- using the SharePoint with [pipeline API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/pipeline-api)
- adding embeddings with [vector API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/vector-api)

We planned to support the grounding [features](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/grounding-3dd1bc414c1848fe869e94902a3ead9f) early next year.

**Comment by zxcvboy** [comment]
Hi. @jjtang1985 

  I want to retrieve some results from my vector engine and send those results with my prompt to LLM.   
  Below is my sample code with Spring AI.  I just wonder if this feature can be achieved by ai-sdk-java or not.
------------ Spring AI sample -------------------------

  private final VectorStore vectorStore; // Spring AI自動配置
  private final ChatClient chatClient; // 使用Config類別建立好Bean,這裡可以直接綁定

 public String search2(String query) throws IOException {

		String result = chatClient.prompt().advisors(
				new QuestionAnswerAdvisor(vectorStore) // RAG增強器
		).user(query).call().content();

		return result;
	}

**Comment by MatKuhr** [comment]
Yes, you can achieve the same RAG flow with the AI SDK. However, currently you have to implement the embedding calls and vector database retrieval manually. We have it on the roadmap to provide an integration for Spring AI, but this isn't completed yet.

You can refer to the [sample code](https://github.com/SAP/ai-sdk-java/blob/main/sample-code/spring-app/src/main/java/com/sap/ai/sdk/app/controllers/OpenAiController.java#L195) on how to compute an embedding.

**Comment by zxcvboy** [comment]
Hi. @MatKuhr.

   Appreciate for your reply. Is there a specific roadmap for integration with Spring AI? RAG is the most popular AI scenarios for SAP customers. Is it possible to provide a sample or scenario for a RAG use case? 

Thanks a lot.

**Comment by MatKuhr** [comment]
Yes, we have actually just released the first iteration of support for Spring AI with version `1.2.0`. You can find the [documentation here](https://github.com/SAP/ai-sdk-java/blob/main/docs/guides/SPRING_AI_INTEGRATION.md). Currently, that is limited to the `OrchestrationClient`, which itself doesn't support embeddings at the moment, but can of course be used together with a vector store. We haven't tested if the current state of the integration already allows for the `QuestionAnswerAdvisor` to work around it, but feel free to try it out and let us know 😄 

Alternatively, one can use the grounding capabilities of orchestration, if you don't want to implement the full RAG flow manually. If that is interesting for you, you can [learn more here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/grounding).

**Comment by newtork** [comment]
(update)

> We planned to support the grounding [features](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/grounding-3dd1bc414c1848fe869e94902a3ead9f) early next year.

FYI, we released `1.4.0` last Friday. It provides initial support for [Document Grounding services](https://github.com/SAP/ai-sdk-java/blob/main/docs/guides/GROUNDING.md). Stable convenience will follow soon.

The Spring AI integration for grounding is still waiting for implementation, but it'll be done in the following weeks hopefully.

**Comment by CharlesDuboisSAP** [comment]
Hello @zxcvboy,

Can you use the Orchestration grounding service, that uses the central HANA database, or, do you have to use your own HANA database?

**Comment by zxcvboy** [comment]
Hi. @CharlesDuboisSAP 

   I tried to leverage grounding capability of Orchestration, however I didn`t find out any document about how to add Vector DB into a resource group. So I am stuck right now.

**Comment by Jonas-Isr** [comment]
Hi @zxcvboy, 

you can find documentation on how to set up the Orchestration vector API [here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/grounding#option-2-provide-the-chunks-of-document-via-vector-api-directly). The steps outlined there are basically:
1. [Create a resource group for grounding](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/create-resource-group-for-ai-data-management)
2. [Preparing Data Using the Vector API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/vector-api)
3. [Configure Grounding Module](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/using-grounding-module)

Does that help?
Best,
Jonas

**Comment by MatKuhr** [comment]
Just as additional hint: In case you want to keep using the `default` resource group, you can also update that using `PATCH` with 

```json
{
	"labels": [   
    {
      "key": "ext.ai.sap.com/document-grounding", 
      "value": "true"
    }
  ]
}
```

**Comment by zxcvboy** [comment]
> Hi [@zxcvboy](https://github.com/zxcvboy),
> 
> you can find documentation on how to set up the Orchestration vector API [here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/grounding#option-2-provide-the-chunks-of-document-via-vector-api-directly). The steps outlined there are basically:
> 
> 1. [Create a resource group for grounding](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/create-resource-group-for-ai-data-management)
> 2. [Preparing Data Using the Vector API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/vector-api)
> 3. [Configure Grounding Module](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/using-grounding-module)
> 
> Does that help? Best, Jonas

Is it possible to do it with AI launchpad? Thanks a lot.

**Comment by newtork** [comment]
Can you please make your question more specific? How is that related to AI SDK?

**Comment by zxcvboy** [comment]
> Can you please make your question more specific? How is that related to AI SDK?

Hi. @newtork.

    Original question is "Is RAG feature supported by ai-sdk-java?".  Thanks a lot.

**Comment by rpanackal** [comment]
Hi @zxcvboy ,

Please refer to the documentation for [AI Launchpad - Grounding Management](https://help.sap.com/docs/ai-launchpad/sap-ai-launchpad/grounding-management?locale=en-US) for the most up to date information on their features.

As far as I know, you can indeed use AI Launched to create a labelled resource group, create data repository (only MSSharePoint) and retrieve data from them. They also feature to use the same together with orchestration. 

Unfortunately, its difficult for us to be on top of the all features they offer. So, I please approach relevant support teams to know more. We are always happy to offer support on usage of our sdk for achieving the same. 

Best regards,
Roshin
